master/internal/proxy/proxy_intg_test.go

```diff
@@ -0,0 +1,107 @@
+package proxy_test
+
+import (
+	"net/url"
+	"testing"
+
+	"github.com/labstack/echo/v4"
+	"github.com/stretchr/testify/require"
+
+	"github.com/determined-ai/determined/master/internal/proxy"
+)
+
+var (
+	proxyAuth = func(c echo.Context) (done bool, err error) {
+		return true, nil
+	}
+	serviceIDs = []string{"a", "b", "c"}
+	u          = url.URL{} // "localhost:8081" TODO CAROLINA
+)
+
+func register(t *testing.T, prTCP bool, unauth bool) {
+	for _, id := range serviceIDs {
+		proxy.Register(id, &u, prTCP, unauth)
+		if proxy.GetService(id) == nil {
+			t.Logf("failed to find registered service %s", id)
+		}
+	}
+	if len(proxy.GetSummary()) != len(serviceIDs) {
+		t.Logf("failed to register all services")
+	}
+}
+
+func unregister(t *testing.T) {
+	for _, id := range serviceIDs {
+		if proxy.GetService(id) == nil {
+			t.Logf("failed to find registered service %s", id)
+		}
+		proxy.Unregister(id)
+		if proxy.GetService(id) != nil {
+			t.Logf("failed to unregister service %s", id)
+		}
+	}
+	if len(proxy.GetSummary()) != 0 {
+		t.Logf("failed to unregister all services.")
+	}
+}
+
+func TestProxyLifecycle(t *testing.T) {
+	cases := []struct {
+		name                 string
+		proxyTCP             bool
+		allowUnauthenticated bool
+	}{
+		// ProxyTCP = AllowUnauthenticated = True
+		{"tcp & unauthenticated true", true, true},
+		// ProxyTCP = AllowUnauthenticated = False
+		{"tcp & unauthenticated false", false, false},
+		// ProxyTCP = true, AllowUnauthenticated = False
+		{"tcp true & unauthenticated false", true, false},
+		// ProxyTCP = False, AllowUnauthenticated = True
+		{"tcp false & unauthenticated true", false, true},
+	}
+
+	// First init the new Proxy
+	proxy.InitProxy(proxyAuth)
+	// And check that the Proxy struct is set up correctly
+	require.NotNil(t, proxy.TheOneProxy.HTTPAuth)
+	require.Equal(t, map[string]*proxy.Service{},
+		proxy.TheOneProxy.Services)
+	require.Equal(t, "", proxy.TheOneProxy.Syslog.Message)
+
+	// Then create the new proxy handler for the services
+	for _, id := range serviceIDs {
+		handler := proxy.NewProxyHandler(id)
```

Comment:
there is 1 proxy handler per cluster, weirdly

Reply:
ooh got it, so it's not per service